### PR TITLE
update to Kokkos 4.0, C++17, and latest VeloC

### PR DESCRIPTION
**Pull request by ElisabethGiem**
Solution to Issue 40, update code to work with newer versions of kokkos develop.

- Updated out of date headers
- Removed calls to verify_is_master() in base OMP instance, as it was removed from updated and renamed Kokkos instance class
- Updated profiling error handling

### CODE REVIEW

File: src/resilience/openMP/OpenMPResParallel.hpp
Comment (dalg24) [inline src/resilience/openMP/OpenMPResParallel.hpp:51]: These are private kokkos headers do not include them

**Comment by nphtan** [comment]
It builds and runs the tests fine but using the VeloC backend (automatic checkpointing) results in errors due to empty memory region selection. It's not capturing and registering the Views with VeloC.

**Comment by nmm0** [comment]
> It builds and runs the tests fine but using the VeloC backend (automatic checkpointing) results in errors due to empty memory region selection. It's not capturing and registering the Views with VeloC.

Perfect -- check out https://github.com/kokkos/kokkos-resilience/blob/main/tests/TestLambdaCapture.cpp#L77

We had to change the API; you have to add the view subscriber to your view types to enable the capture.

File: src/resilience/openMP/OpenMPResParallel.hpp
Comment (ElisabethGiem) [reply src/resilience/openMP/OpenMPResParallel.hpp:51]: Sorry for the late reply, I was out for a while due to illness. They have been removed, and I'm in the process of purging all private headers now that I have the list.

**Review by nmm0 — CHANGES_REQUESTED** [review]
Develop changed again, making more private headers cause compiler failures. We also need to audit our existing code for this.

**Comment by Matthew-Whitlock** [comment]
These changes seem to also update to work with version of VeloC in your VeloC PR 43, was that intentional? I'm reviewing with the assumption that it was, though it might be good to wait until that is merged in.

1.  **Major**: cmake/resilienceConfig.cmake.in should be updated to the new syntax for veloc (`find_dependency(veloc)` instead of `find_dependency(VeloC)`)
2. **Minor**: Since this requires using the newest version of VeloC, and VeloC barebones is not being maintained, we should remove references to it in CMakeLists, CMake config, cmake presets, and readme. The readme mentions spack options for compiling VeloC to barebones - does the spack stuff also need to be updated? 


The test framework is failing in two ways, but they seem unrelated to these changes. Just listing here for reference so no one else needs to track down the sources when testing this:
-   A misleading failure of `TestResOpenMP.TestResilientForInsertError` caused by a multiple of three number of OpenMP threads.  I think we could use omp_get/set_num_threads to make this more reliable.
-  Failures on both TestVelocMemoryBackend and TestStdFileBackend, due to a failure to enter checkpoint or recovery. This is due to a potentially overprotective ifdef guard: https://github.com/kokkos/kokkos-resilience/blob/40-update-code-to-work-with-newer-versions-of-kokkos-develop/src/resilience/AutomaticCheckpoint.hpp#L82

Other than these, I think the code looks good.

File: CMakeLists.txt
Comment (dalg24) [inline CMakeLists.txt:15]:
```suggestion
if (KR_ALL_WARNINGS)
```

File: CMakeLists.txt
Comment (dalg24) [inline CMakeLists.txt:13]: Do you really want an option for these?  `-DCMAKE_CXX_FLAGS="-Wall -Werror"` is easy enough IMO.

File: CMakeLists.txt
Comment (dalg24) [inline CMakeLists.txt:36]:
```suggestion
```
would never occur as you require a minimum version of 4.0

File: CMakeLists.txt
Comment (dalg24) [inline CMakeLists.txt:52]: Probably should propose these VeloC changes as part as another PR

File: CMakeLists.txt
Comment (dalg24) [reply CMakeLists.txt:13]: Also has nothing to do with the bumping Kokkos to 4.0

**Comment by Matthew-Whitlock** [comment]
https://github.com/kokkos/kokkos-resilience/blob/main/src/resilience/AutomaticCheckpoint.hpp#L82

Should be updated to use KOKKOS_IF_ON_HOST

File: CMakeLists.txt
Comment (nmm0) [reply CMakeLists.txt:13]: That's true, but it's a good way to avoid compiler differences. CMake even added a property in more recent versions for Werror. As far as Kokkos 4.0 updates sadly you are correct. It's just we had... a lot of problems with the current way the CMake was set up and not making warnings not errors by default (previous behavior) solves them

File: CMakeLists.txt
Comment (nmm0) [reply CMakeLists.txt:52]: VeloC is changing the way their cmake works (writing a config file). It's a bit messy but maybe the PR should be titled "Update code to work with newer versions of Kokkos and VeloC develop"

**Comment by nmm0** [comment]
Well the already bloated PR got a bit more bloated. I also changed some usages of `boost::optional` and `boost::variant` to `std::optional` and `std::variant`

**Comment by nmm0** [comment]
> Matthew's major concern from the previous review (cmake/resilienceConfig.cmake.in should be updated to the new syntax for veloc ) was not addressed in the files that came with this branch for me. Other than that, the new commits seem fine to me.

Woops, forgot that, fixed now!

File: src/resilience/Config.hpp
Comment (Matthew-Whitlock) [inline src/resilience/Config.hpp:199]: I'm getting an error here in the map_type, which uses a key/val tuple that includes Entry. 
```
stl_pair.h: In instantiation of 'struct std::pair<const std::__cxx11::basic_string<char>, KokkosResilience::Config::Entry>'
```
Entry isn't yet fully defined.
```
stl_pair.h:218:11: error: 'std::pair<_T1, _T2>::second' has incomplete type
```
Probably solved by making this a unique pointer to the variant?

**Review by Matthew-Whitlock — CHANGES_REQUESTED** [review]
One blocking change, though another minor fix would be to remove the src/resilience/filesystem/Filesystem.*pp files and switch to just using std::filesystem in the two tests that use it.

File: src/resilience/Config.hpp
Comment (nmm0) [reply src/resilience/Config.hpp:199]: Fixed by making it a regular map, unfortunately the value has to be copied so unique_ptr doesn't work here

**Review by Matthew-Whitlock — APPROVED** [review]
Compiles great, builds and tests successfully against latest Kokkos, looks good!

**Pull request merged**
